include/pybind11/numpy.h

```diff
@@ -108,6 +108,15 @@ inline numpy_internals& get_numpy_internals() {
     return *ptr;
 }
 
+// Lookup a type according to its size, and return a value corresponding to the NumPy typenum.
+template <typename Concrete,
+    typename Check1, typename Check2, typename Check3 = Check2>
+constexpr int platform_lookup(int check1, int check2, int check3 = -1) {
+    return (sizeof(Concrete) == sizeof(Check1)) ? check1 :
+               (sizeof(Concrete) == sizeof(Check2)) ? check2 :
+                 (sizeof(Concrete) == sizeof(Check3)) ? check3 : -1;
+}
```

Comment:
Here's a template that makes this a little more generic, along with a compile-time check for the error case (rather than returning -1).  It also requires passing the arguments as an `{ ... }` array because that seemed simpler.
```C++
using namespace pybind11::detail;

template <typename T> struct same_size { template <typename U> using as = bool_constant<sizeof(T) == sizeof(U)>; };

template <typename Concrete, typename... Check>
constexpr int platform_lookup(std::array<int, sizeof...(Check)> codes) {
    using code_index = std::integral_constant<int, constexpr_first<same_size<Concrete>::template as, Check...>()>;
    static_assert(code_index::value != sizeof...(Check), "Unable to match type on this platform");
    return codes[code_index::value];
}

int main() {
    // fails (short not matched):
    //int bar = platform_lookup<short, int, unsigned, unsigned long, int>({4, 9, 11, 13});
    int foo = platform_lookup<std::uint32_t, unsigned int, unsigned long>({8,9});
}
```

Reply:
Unfortunately, C++11 doesn't like multiple statements in the function - it deems it non-`constexpr` :(

Will briefly tinker with it to see if I can add some indirection to preserve the `constexpr`-ness.

Reply:
Hup, was wrong: Just needed to make the `array<>` argument `const array<>`.
Thanks!